### ISSUE TITLE
Interface with the extension type from cython

### ISSUE THREAD

**Issue by jorisvandenbossche** [caspervdw/pygeos#24]
As an experiment, I was trying to implement something in cython using the extension type / object array holding them (a simple vectorized contains; not to replace the contains ufunc, but as simple example of how to use cython with an array of pygeos objects, for potentially a custom non-ufunc algorithm).

So I succeeded in interfacing with the GeometryObject struct (to get the pointer out of the struct), without Python interaction. 

But what I didn't get to work yet is to access a single pygeos.Geometry object out of the object array (which needs to happen in the loop). 
And the result of this is that I also can't put this for loop in a `nogil` block, which would be needed to run it in a parallel context.

See experiment here: https://nbviewer.jupyter.org/gist/jorisvandenbossche/c90181a330972d5f4564d7c73b70a114

**Comment by caspervdw** [comment]
Great that you got it to work with Cython! 

I think the numpy-accessing bit is an issue unrelated to PyGEOS. Maybe you need to type the index correctly?

**Comment by jorisvandenbossche** [comment]
It's not the index I think (that one is typed, and you can also see it in the output of the annotated code). As far as I can see, the problem is that you get a pygeos.Geometry object out of the array, and cast that to a GeometryObject. That casting to a GeometryObject is fine, as that is something that cython knows about, but getting the object out as pygeos.Geometry is the problem.

The annotated code shows (the first line is the cython line, the lines below that the C code that is generated from it):

```
+56:         pygeom2 = <GeometryObject *> array[idx]

    __pyx_t_10 = __pyx_v_idx;
    __pyx_t_6 = (PyObject *) *__Pyx_BufPtrStrided1d(PyObject **, __pyx_pybuffernd_array.rcbuffer->pybuffer.buf, __pyx_t_10, __pyx_pybuffernd_array.diminfo[0].strides);
    __Pyx_INCREF((PyObject*)__pyx_t_6);
    __pyx_v_pygeom2 = ((GeometryObject *)__pyx_t_6);
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;

```

So to cast the `array[idx]` element to a GeometryObject and assign it to `pygeom2` variable, cython creates a temporary PyObject. 
And I am not sure how to tell cython (or whether this is possible) that the object array holds all GeometryObjects, so that it can avoid this temporary python object.

**Comment by caspervdw** [comment]
Right, would this mean that Cython casts any object from an object-typed bumpy array?

Btw the GeometryObject is the C symbol I use for an instance of pygeos.Geometry.

**Comment by caspervdw** [comment]
This issue is probably the core problem of #226 .
